doc/design/record_fetcher_proposal.md

```diff
@@ -0,0 +1,178 @@
+opinions regarding how things are stored, but we need a specific API to ensure safe collection. Our collection API:
+* must allow writing new documents and manifests
+* must not allow for the fetching or altering of existing documents or manifests
+* must enforce authentication and authorization for document submission
+  * this lets us prevent unknown/untrusted entities from aiming a firehose of data at us and eating all of our money,
+  but there's little risk of this threatening the trustworthiness of our records due to the lack of manifest
+  * this document doesn't have a strong opinion here regarding how we authenticate, but the appendix has a strawman we
+can use to start that conversation
+* must not alter the manifest or its signature before committing to durable storage
+  * without the signature, we can't detect tampering or associate the manifest with a fetcher
+* should probably losslessly compress most things (we get 40% savings with gzip on the sample manifest below)
+    
+## Design Ideas
+### Document Storage
+First, let's talk document storage. Have record fetchers crawl however they want and upload documents to a document
+storage endpoint. This should ideally be some kind of content-addressed blob storage service. We don't need a lot of
+structure for this, as any relationship between documents will be encoded by the manifests -- the service just needs to
+be able to store the raw documents for later lookup. Backing this storage API with a fairly flat filesystem is fine
+here, we're rarely going to list files, only write and read. If we do need to list files, it'll be for some cleanup
+operation and should not be on any request path.
+
+A big benefit of content-addressed storage here is that we get deduping almost for free -- we may need to zero out
+content that changes on every page load (e.g. "This page requested at HH:MM"), but that's more of an optimization than a
+hard requrement. These savings would apply to both documents which remain unchanged between fetches and documents that
+are shared between records.
+
+The actual document collection service will take a file upload and respond with a simple payload denoting the URI that
+may be used to access the file.
+
+### Manifest Construction
+
+Manifests need to be signed, so we'll just use [RFC 7519](https://tools.ietf.org/html/rfc7519) JWT blobs, encoding the
+manifest's information in the JWT payload. As an example, the payload could resemble:
+
+```json
+{
+  "version": 1,
+  "uuid": "6e566397-6803-49ec-9280-62c00b368e3f",
+  "record_type": "USA/FL/Bay/Court",
+  "case_id": "20000001",
+  "collected_at_ms": 1592618026000,
+  "fetcher_id": "github.com/Police-Data-Accessibility-Project/Scrapers:USA/FL/Bay/Court/scraper/Scraper.py@1db20ce1e63768363d532d46a641722986f63524",
+  "documents": [
+    {
+      "label": "some pdf we'll have to ocr",
+      "content_type": "application/pdf",
+      "digest": "...",
+      "uri": "https://..."
```

Comment:
This makes it seem like every document is stored individually?  I was assuming the "record" was a bundle that's atomically stored as a unit. It would provide easier access patterns (you fetch and save everything together) for instance.  Either all of it would be stored or none of it, avoiding issues of partially uploaded data.

It should be easy to manage bundles using MIME Multipart, which is widely supported in various languages too
https://en.wikipedia.org/wiki/MIME#Multipart_messages

Is there an advantage to keeping everything separate?

Reply:
I assumed a "record" was a specific event or case that my span multiple retrieved documents, storing the documents with an audit trail and the auditably constructing the "record" from source would give a chain of custody and allow anyone to reproduce the record from the source material.

Reply:
> This makes it seem like every document is stored individually?

That was the intent.

>  I was assuming the "record" was a bundle that's atomically stored as a unit.

That is a *way* better idea! Completely removes the garbage collection requirement by preventing the scenario from ever arising.